File: solutions/elixir/roman-numerals/2/lib/roman_numerals.ex
defmodule RomanNumerals do
  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
    numeral(number, [])
    |> Enum.reverse()
    |> IO.iodata_to_binary()
  end

  defp numeral(0, numeral), do: numeral
  defp numeral(number, numeral) when number >= 1000, do: numeral(number - 1000, [?M | numeral])
  defp numeral(number, numeral) when number >= 900, do: numeral(number - 900, [?M, ?C | numeral])
  defp numeral(number, numeral) when number >= 500, do: numeral(number - 500, [?D | numeral])
  defp numeral(number, numeral) when number >= 400, do: numeral(number - 400, [?D, ?C | numeral])
  defp numeral(number, numeral) when number >= 100, do: numeral(number - 100, [?C | numeral])
  defp numeral(number, numeral) when number >= 90, do: numeral(number -  90, [?C, ?X | numeral])
  defp numeral(number, numeral) when number >= 50, do: numeral(number -  50, [?L | numeral])
  defp numeral(number, numeral) when number >= 40, do: numeral(number -  40, [?L, ?X | numeral])
  defp numeral(number, numeral) when number >= 10, do: numeral(number -  10, [?X | numeral])
  defp numeral(number, numeral) when number >= 9, do: numeral(number -  9, [?X, ?I | numeral])
  defp numeral(number, numeral) when number >= 5, do: numeral(number -  5, [?V | numeral])
  defp numeral(number, numeral) when number >= 4, do: numeral(number -  4, [?V, ?I | numeral])
  defp numeral(number, numeral) when number >= 1, do: numeral(number -  1, [?I | numeral])
end

File: solutions/erlang/roman-numerals/1/src/roman_numerals.erl
-module(roman_numerals).

-export([roman/1]).


roman(N) -> roman(N, []).

roman(N, Roman) when N >= 1000 -> roman(N - 1000, [$M | Roman]);
roman(N, Roman) when N >= 900  -> roman(N - 900, [$M, $C | Roman]);
roman(N, Roman) when N >= 500  -> roman(N - 500, [$D | Roman]);
roman(N, Roman) when N >= 400  -> roman(N - 400, [$D, $C | Roman]);
roman(N, Roman) when N >= 100  -> roman(N - 100, [$C | Roman]);
roman(N, Roman) when N >= 90   -> roman(N - 90, [$C, $X | Roman]);
roman(N, Roman) when N >= 50   -> roman(N - 50, [$L | Roman]);
roman(N, Roman) when N >= 40   -> roman(N - 40, [$L, $X | Roman]);
roman(N, Roman) when N >= 10   -> roman(N - 10, [$X | Roman]);
roman(9, Roman) -> roman(0, [$X, $I | Roman]);
roman(N, Roman) when N >= 5    -> roman(N - 5, [$V | Roman]);
roman(4, Roman) -> roman(0, [$V, $I | Roman]);
roman(3, Roman) -> roman(0, [$I, $I, $I | Roman]);
roman(2, Roman) -> roman(0, [$I, $I | Roman]);
roman(1, Roman) -> roman(0, [$I | Roman]);
roman(0, Roman) -> string:reverse(Roman).
